generator/src/main/java/com/scottlogic/deg/generator/generation/RegexStringGenerator.java

```diff
@@ -424,35 +422,89 @@ void setMaxChar(char maxChar) {
     private class FiniteStringAutomatonIterator implements Iterator<String> {
 
         private final RegexStringGenerator stringGenerator;
+        private final long matches;
         private int currentIndex;
+        private String currentValue;
 
         FiniteStringAutomatonIterator(RegexStringGenerator stringGenerator) {
             this.stringGenerator = stringGenerator;
+            this.matches = stringGenerator.getValueCount();
             currentIndex = 0;
         }
 
+        /**
+         * <p>
+         * This function has been updated to only allow valid single 16-bit
+         * word UTF-8 characters to be output.
+         * </p>
+         * <p>
+         *     FIXME - This check will be removed if/when the dk.brics.automaton
+         *     library is fixed to support surrogate pairs,
+         * </p>
+         * <p>
+         *     issue #15 (https://github.com/cs-au-dk/dk.brics.automaton/issues/15)
+         *     has been raised on the dk.brics.automaton library
+         * </p>
+         * <p>
+         *     issue #537 has been created to track when the dk.brics.automaton library
+         *     is updated.
+         * </p>
+         */
         @Override
         public boolean hasNext() {
-            long matches = stringGenerator.getValueCount();
-            return currentIndex < matches;
+            if (currentValue != null) {
+                return true;
+            }
+            do {
+                currentIndex++; // starts at 1
```

Comment:
Do we not check the 0th index?

Reply:
And the current index does actually start at 0, the comment is misleading

Reply:
well, although the variable is initialised to 0, the first value we check is after the increment, so the first value we check is 1.
Why?, I don't know. that was the code before my change

Reply:
Is this a requirement of the automaton, can you check in case we've found a bug?
Maybe the comment can be updated to clarify that the automaton works on 1-based indexes if that is the case?